### ISSUE TITLE
Reverse operators don't follow the subclassing rule.

### ISSUE THREAD

**Issue by Dr-Irv** [microsoft/pyright#4088]
**Describe the bug**
From the python docs at https://docs.python.org/3/reference/datamodel.html?highlight=__add__#object.__add__  :
" If the right operand’s type is a subclass of the left operand’s type and that subclass provides a different implementation of the reflected method for the operation, this method will be called before the left operand’s non-reflected method. This behavior allows subclasses to override their ancestors’ operations."

This is not obeyed by pyright (or mypy)

**To Reproduce**
```python
from typing_extensions import reveal_type


class Base:
    def __init__(self):
        ...

    def __add__(self, other: Base) -> Base:
        return Base()

    def __radd__(self, other: Base) -> Base:
        return Base()


class Sub(Base):
    def __add__(self, other: Sub) -> Sub:
        return Sub()

    def __radd__(self, other: Sub) -> Sub:
        return Sub()


b = Base()
s = Sub()

ad = b + s
reveal_type(ad)
print(type(ad))
```

runtime output:
```text
Runtime type is 'Sub'
<class '__main__.Sub'>
```
pyright output:
```text
addsc.py:30:13 - information: Type of "ad" is "Base"
```

**Expected behavior**
The revealed type (as shown at runtime) should be `Sub`

**Screenshots or Code**

See above.

**VS Code extension or command-line**

pyright command line 1.1.276

**Additional context**

Came up when doing something similar with `timedelta` as the base class and `pandas.TimeDelta` as the subclass.

**Comment by erictraut** [comment]
This has been discussed many times before. I don't see a good solution to this problem.

In the Python type system, a symbol that is declared as type `Foo` means that it's a subclass of `Foo` or a subclass thereof. There's no way to know statically whether it's actually a `Foo` or a subclass. That information is required to determine which dynamic code path will be used at runtime when evaluating an operator.

For this reason, the current behavior is "as designed", and we don't plan to make any further changes here.
